### PR TITLE
Update Cookie Auth without Identity topic

### PR DESCRIPTION
**Pull request by guardrex**
Addresses #3777

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/security/authentication/cookie?branch=pr-en-us-4538)

#### Adds a sample to demo cookie auth without identity
* This is a Razor Pages sample. :tada: 🎈 :beers: ... it even includes an `AuthorizePage` convention to show how user auth can permit/deny page access.
* No implementation for validation of user credentials. It's very simple in that regard, since it's only trying to show how to setup the app for cookie auth. Devs will plug in a thousand things there, so I punt to them at that point.
* This is the first example of a sample app with full static assets and no Bower/`BundlerMinifier.Core` bits. **Also note:** Due to lifting the restriction on *lib* folder assets, this PR puts them in for the Razor Pages tutorial app, which should be fine.

#### Topic highlights
* Performed the usual updates for guidelines adherence.
* Adds a desperately needed opening paragraph on why u would even ever do this.
* Updates options
* I think this is the first time the docs surface Cookie Policy Middleware. :tada:
* Big attempt here to link API topics.

cc/ @JunTaoLuo @Tratcher @HaoK @blowdart @PinpointTownes

### CODE REVIEW

File: aspnetcore/security/authentication/cookie.md [hunk 1]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+                           select c.Value).FirstOrDefault();
 
-        if (string.IsNullOrEmpty(lastChanged) ||
-            !userRepository.ValidateLastChanged(userPrincipal, lastChanged))
+        if (string.IsNullOrEmpty(lastUpdated) ||
+            !userRepository.ValidateLastUpdated(userPrincipal, lastUpdated))
         {
             context.RejectPrincipal();
-            await context.HttpContext.SignOutAsync("MyCookieAuthenticationScheme");
+            await context.HttpContext.SignOutAsync(
```
Comment (PinpointTownes) [inline aspnetcore/security/authentication/cookie.md]: @guardrex `context.HttpContext.SignOutAsync()` -> `context.SignOutAsync()`.

File: aspnetcore/security/authentication/cookie.md [hunk 2]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
-This would then be wired up during cookie service registration in the `ConfigureServices` method of *Startup.cs*:
+Register the event during cookie service registration in the `ConfigureServices` method:
 
 ```csharp
-services.AddAuthentication("MyCookieAuthenticationScheme")
-        .AddCookie("MyCookieAuthenticationScheme", options =>
+services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
+    .AddCookie(options =>
+    {
+        options.Events = new CookieAuthenticationEvents
```
Comment (PinpointTownes) [inline aspnetcore/security/authentication/cookie.md]: It's worth noting that you can now use DI with the events classes in 2.0. Maybe that would a good opportunity to show that with the validator thingy.

Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: I agree. I'll update this later today. Thanks for the tip. Constructor injection makes a lot more sense than `GetRequiredService` ... ei ei ei *yuck!*

File: aspnetcore/security/authentication/cookie.md [hunk 3]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+| [SessionStore](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.sessionstore?view=aspnetcore-2.0) | An optional container used to store identity across requests. When used, only a session identifier is sent to the client. `SessionStore` can be used to mitigate potential problems with large identities. |
+| [SlidingExpiration](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.slidingexpiration?view=aspnetcore-2.0) | A flag indicating whether the cookie expiration date resets when more than half of the `ExpireTimeSpan` interval has passed. The new expiration date is moved forward to be the current date plus the `ExpireTimespan`. An [absolute cookie expiration time](xref:security/authentication/cookie#absolute-cookie-expiration) can be set by using the `AuthenticationProperties` class when calling `SignInAsync`. An absolute expiration time can improve the security of your app by limiting the amount of time that the authentication cookie is valid. The default value is `true`. |
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: indicating if a new cookie with updated expiration should be issued dynamically. This can happen on any request where the current cookie is more than 50% expired.

File: aspnetcore/security/authentication/cookie.md [hunk 4]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+| [Cookie.SameSite](/dotnet/api/microsoft.aspnetcore.http.cookiebuilder.samesite?view=aspnetcore-2.0) | Indicates whether the browser should allow the cookie to be attached to same-site requests only (`SameSiteMode.Strict`) or cross-site requests using safe HTTP methods and same-site requests (`SameSiteMode.Lax`). When set to `SameSiteMode.None`, cookies are used for same-site and cross-site requests if the [MinimumSameSitePolicy](/dotnet/api/microsoft.aspnetcore.builder.cookiepolicyoptions.minimumsamesitepolicy) is also `SameSiteMode.None`; otherwise, the value for `Cookie.SameSite` is taken from `MinimumSameSitePolicy` (see [Cookie Policy Middleware](#cookie-policy-middleware) below). To support OAuth authentication, the default value is `SameSiteMode.Lax`. For more information, see [OAuth authentication broken due to SameSite cookie policy](https://github.com/aspnet/Security/issues/1231). |
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: Don't discuss cookiepolicyoptions here. That's an independent topic.

File: aspnetcore/security/authentication/cookie.md [hunk 5]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
-# [ASP.NET Core 1.x](#tab/aspnetcore1x)
+**AddCookie Options**
 
-Complete the following steps:
+The [CookieAuthenticationOptions](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions?view=aspnetcore-2.0) class is used to configure the authentication provider options.
 
-- Install the `Microsoft.AspNetCore.Authentication.Cookies` NuGet package in your project. This package contains the cookie middleware.
+| Option | Description |
+| ------ | ----------- |
+| [AccessDeniedPath](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.accessdeniedpath?view=aspnetcore-2.0) | Informs the handler that it should change an outgoing 403 Forbidden status code into a 302 Found (URL redirect) with the provided path. The default value is `/Account/AccessDenied`. |
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: This does not react to 403s, it is only triggered by HttpContext.ForbidAsync().

File: aspnetcore/security/authentication/cookie.md [hunk 6]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+| [ExpireTimeSpan](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.expiretimespan?view=aspnetcore-2.0) | The `TimeSpan` after which the authentication ticket stored inside the cookie expires. `ExpireTimeSpan` is added to the current time to create the expiration time for the ticket. To use `ExpireTimeSpan`, you must set `IsPersistent` to `true` in the `AuthenticationProperties` passed to `SignInAsync`. The default value is 14 days. |
+| [LoginPath](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.loginpath?view=aspnetcore-2.0) | Informs the handler that it should change an outgoing 401 Unauthorized status code into a 302 Found (URL redirect) with the provided path. The current URL that generated the 401 is added to the `LoginPath` as a query string parameter named by the `ReturnUrlParameter`. Once a request to the `LoginPath` grants a new sign-in identity, the `ReturnUrlParameter` value is used to redirect the browser back to the URL that caused the original unauthorized status code. The default value is `/Account/Login`. |
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: This does not react to 401s, it reacts to HttpContext.ChallengeAsync()

File: aspnetcore/security/authentication/cookie.md [hunk 7]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+| [LogoutPath](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.logoutpath?view=aspnetcore-2.0) | If the `LogoutPath` is provided to the handler, then a request to that path redirects based on the value of the `ReturnUrlParameter`. The default value is `/Account/Logout`. |
+| [ReturnUrlParameter](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.returnurlparameter?view=aspnetcore-2.0) | Determines the name of the query string parameter that is appended by the handler when a 401 Unauthorized status code is changed to a 302 Found (URL redirect) to the `LoginPath`. `ReturnUrlParameter` is used when a request arrives on the `LoginPath` or `LogoutPath` to return the browser to the original URL after the login or logout action is performed. The default value is `ReturnUrl`. |
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: remove references to 401

**Comment by guardrex** [comment]
:smile: ... that seems like a considerable improvement! Thx @PinpointTownes! :beer:

@HaoK When you get to this, check my link for `ValidateAsync` for the 1.x example of *Reacting to back-end changes*. I just want to make sure I have the right destination :point_right: [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.isecuritystampvalidator.validateasync?view=aspnetcore-1.1).

File: aspnetcore/security/authentication/cookie.md [hunk 8]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+
+---
 
-* `LoginPath` - This is the relative path to which requests redirect when a user attempts to access a resource but has not been authenticated.
+## Cookie Policy Middleware
 
-[Other options](xref:security/authentication/cookie#security-authentication-cookie-options) include the ability to set the issuer for any claims the cookie authentication creates, the name of the cookie the authentication drops, the domain for the cookie and various security properties on the cookie. By default, the cookie authentication uses appropriate security options for any cookies it creates, such as:
-- Setting the HttpOnly flag to prevent cookie access in client-side JavaScript
-- Limiting the cookie to HTTPS if a request has traveled over HTTPS
+[Cookie Policy Middleware](/dotnet/api/microsoft.aspnetcore.cookiepolicy.cookiepolicymiddleware) enables cookie policy capabilities in an app:
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: Note that this is order sensitive, it only affects components registered after it in the pipeline.

**Comment by PinpointTownes** [comment]
> :smile: ... that seems like a considerable improvement! Thx @PinpointTownes! :beer:

My pleasure :call_me_hand:

File: aspnetcore/security/authentication/cookie.md [hunk 9]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
-await HttpContext.Authentication.SignInAsync("MyCookieAuthenticationScheme", principal);
+await HttpContext.Authentication.SignInAsync(
+    CookieAuthenticationDefaults.AuthenticationScheme,
+    new ClaimsPrincipal(claimsIdentity));
 ```
 
 ---
 
-This creates an encrypted cookie and adds it to the current response. The `AuthenticationScheme` specified during [configuration](xref:security/authentication/cookie#security-authentication-cookie-middleware-configuring) must be used when calling `SignInAsync`.
+`SignInAsync` creates an encrypted cookie and adds it to the current response. The `AuthenticationScheme` specified during [configuration](xref:security/authentication/cookie#configuration) must be used when calling `SignInAsync`.
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: Or else the default scheme will be used.

**Comment by HaoK** [comment]
@guardrex yup that link to identity looks good

File: aspnetcore/security/authentication/cookie.md [hunk 10]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+
 ```csharp
-await HttpContext.Authentication.SignOutAsync("MyCookieAuthenticationScheme");
+await HttpContext.Authentication.SignOutAsync(
+    CookieAuthenticationDefaults.AuthenticationScheme);
 ```
 
 ---
 
+If you aren't using `CookieAuthenticationDefaults.AuthenticationScheme` (or "Cookies") as the scheme (for example, "ContosoCookie"), supply the scheme you used when configuring the authentication provider.
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: Or else the default will be used.

**Comment by HaoK** [comment]
Nit: The wording in the first paragraph is a bit awkward: "You can use ASP.NET Core Authentication Middleware for cookie-based authentication"

Maybe shorten this to just "You can use cookie-based authentication as a standalone authentication provider without ASP.NET Core Identity"?

File: aspnetcore/security/authentication/cookie.md [hunk 11]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+
 ## Reacting to back-end changes
 
->[!WARNING]
-> Once a principal cookie has been created, it becomes the single source of identity. Even if you disable a user in your back-end systems, the cookie authentication has no knowledge of this, and a user stays logged in as long as their cookie is valid.
+Once a cookie is created, it becomes the single source of identity. Even if you disable a user in your back-end systems, the cookie authentication system has no knowledge of this, and a user stays logged in as long as their cookie is valid.
 
-The cookie authentication provides a series of events in its option class. The `ValidateAsync()` event can be used to intercept and override validation of the cookie identity.
+The `ValidateAsync()` event can be used to intercept and override validation of the cookie identity. This approach mitigates the risk of unauthorized access to an app.
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: mitigates the risk of revoked users accessing the application.

File: aspnetcore/security/authentication/cookie.md [hunk 12]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
 
-Consider a back-end user database that may have a "LastChanged" column. In order to invalidate a cookie when the database changes, you should first, when [creating the cookie](xref:security/authentication/cookie#security-authentication-cookie-middleware-creating-a-cookie), add a "LastChanged" claim containing the current value. When the database changes, the "LastChanged" value should be updated.
+One approach to cookie validation is based on keeping track of when the user database has been changed. If the database hasn't been changed since the user's cookie was issued, there is no need to re-authenticate the user if their cookie is still valid. To implement this scenario, the database stores a `LastUpdated` value. When any user is updated in the database, the `LastUpdated` value is set to the current time.
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: @HaoK

**Comment by guardrex** [comment]
@HaoK Yes, that's much better. I'll wait and batch your changes on the next commit after ur done looking it over. No rush ... I have *plenty* of things to keep me busy. 😅

File: aspnetcore/security/authentication/cookie.md (hunk 12 above)
Comment (HaoK) [reply aspnetcore/security/authentication/cookie.md]: This is reasonable, identity's security stamp works a bit differently, but this should work too. Basic idea is you just need to check a claim against something in the database that changes when you want to invalidate the user's cookie.  LastUpdated seems fine as a choice, (identity generates a new guids for the stamps)

File: aspnetcore/security/authentication/cookie.md [hunk 13]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
-- If not using the `Microsoft.AspNetCore.All` [metapackage](xref:fundamentals/metapackage), install version 2.0+ of the `Microsoft.AspNetCore.Authentication.Cookies` NuGet package in your project.
+In the `ConfigureServices` method, create the Authentication Middleware service with the `AddAuthentication` and `AddCookie` methods:
 
-- Invoke the `UseAuthentication` method in the `Configure` method of the *Startup.cs* file:
+[!code-csharp[Main](cookie/sample/Startup.cs?name=snippet1)]
 
-    ```csharp
-    app.UseAuthentication();
-    ```
+`AuthenticationScheme` passed to `AddAuthentication` sets the authentication scheme. `AuthenticationScheme` is useful when there are multiple instances of cookie authentication and you want to [limit authorization to one instance](xref:security/authorization/limitingidentitybyscheme). Setting the `AuthenticationScheme` to `CookieAuthenticationDefaults.AuthenticationScheme` provides a value of "Cookies" for the scheme. You can supply any string value that distinguishes the scheme.
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: I didn't think that your interpretation was what the original author intended to be taken for its meaning, right? I think they're just pointing out that *later* when you want to limit an auth method call (e.g., signin or signout) to a particular instance that this feature has value. Let's remove that line. Let's put a link to *Limiting identity by scheme* in the *See also* section.

File: aspnetcore/security/authentication/cookie.md (hunk 5 above)
Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: hum ... this isn't good then :point_right: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.accessdeniedpath?view=aspnetcore-2.0

File: aspnetcore/security/authentication/cookie.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: I don't think that `SameSiteMode.None` can be mentioned unless Cookie Policy Middleware is brought in because they interact to set this same-site value. If I pull all of that (or perhaps just move it down to the new Cookie Policy Middleware section), what am I supposed to say about `SameSiteMode.None` here ... just that it doesn't set a `Cookie.SameSite` value?

File: aspnetcore/security/authentication/cookie.md [hunk 14]
```diff
@@ -14,335 +14,369 @@ uid: security/authentication/cookie
...
+| [Events](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.events?view=aspnetcore-2.0) | The handler calls methods on the provider that give the app control at certain processing points. If `Events` aren't provided, a default instance is supplied that does nothing when the methods are called. |
+| [EventsType](/dotnet/api/microsoft.aspnetcore.authentication.authenticationschemeoptions.eventstype?view=aspnetcore-2.0) | Used as the service type to get the `Events` instance instead of the property. |
+| [ExpireTimeSpan](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.expiretimespan?view=aspnetcore-2.0) | The `TimeSpan` after which the authentication ticket stored inside the cookie expires. `ExpireTimeSpan` is added to the current time to create the expiration time for the ticket. To use `ExpireTimeSpan`, you must set `IsPersistent` to `true` in the `AuthenticationProperties` passed to `SignInAsync`. The default value is 14 days. |
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: I'm taking another crack at it along those lines with slightly different (and some stolen 🚓) language from you. Let's see on the next commit how it goes.

File: aspnetcore/security/authentication/cookie.md (hunk 6 above)
Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: Are you sure? :grinning: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationoptions.loginpath?view=aspnetcore-2.0 ... The API Browser has been telling me some nasty fibs lately. lol

@Rick-Anderson Are the API topic updates to be done on our side? If so, I'll open an issue with these that @Tratcher has identified.

**Comment by guardrex** [comment]
Thanks @Tratcher 

I'm still a bit unclear on ...

> Don't discuss cookiepolicyoptions here. That's an independent topic.

... because without mentioning it I'm not sure what to say about `SameSiteMode.None`, unless it's enough to just say ...

> A value of \`SameSiteMode.None\` doesn't set a \`Cookie.SameSite\` value.

File: aspnetcore/security/authentication/cookie.md (hunk 4 above)
Comment (Tratcher) [reply aspnetcore/security/authentication/cookie.md]: Right, SameSiteMode.None does not set any value in the cookie header. It's the opt-out.

File: aspnetcore/security/authentication/cookie.md (hunk 6 above)
Comment (Tratcher) [reply aspnetcore/security/authentication/cookie.md]: Hmm, looks like the doc comments didn't get updated.
https://github.com/aspnet/Security/blob/bd07f8b683ce793490d108b2310fa6112953d172/src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationOptions.cs#L83
Filed: https://github.com/aspnet/Security/issues/1483

**Comment by guardrex** [comment]
@Tratcher The last commit simplifies the remark in `Cookie.SameSite` in favor of a matrix in the Cookie Policy Middleware section.

File: aspnetcore/security/authentication/cookie.md [hunk 15]
```diff
@@ -210,17 +210,17 @@ If you aren't using `CookieAuthenticationDefaults.AuthenticationScheme` (or "Coo
 
 Once a cookie is created, it becomes the single source of identity. Even if you disable a user in your back-end systems, the cookie authentication system has no knowledge of this, and a user stays logged in as long as their cookie is valid.
 
-The `ValidateAsync()` event can be used to intercept and override validation of the cookie identity. This approach mitigates the risk of unauthorized access to an app.
+The [ValidatePrincipal](/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationevents.validateprincipal) event in ASP.NET Core 2.x or the [ValidateAsync](/dotnet/api/microsoft.aspnetcore.identity.isecuritystampvalidator.validateasync) method in ASP.NET Core 1.x can be used to intercept and override validation of the cookie identity. This approach mitigates the risk of unauthorized access to an app.
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: ValidatePrincipal is the same method in 1.x and 2.x. ISecurityStampValidator is called inside of Identity's implementation of ValidatePrincipal.

Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: It's `ValidatePrincipal` vs. `ValidateAsync` tho, which reflects that the 2.x sample is using `ValidatePrincipal`, while I left the 1.x sample using the `ValidateAsync` unchanged.

File: aspnetcore/security/authentication/cookie.md [hunk 16]
```diff
@@ -14,202 +14,332 @@ uid: security/authentication/cookie
...
-        {
-            options.Events = new CookieAuthenticationEvents
-            {
-                OnValidatePrincipal = LastChangedValidator.ValidateAsync
-            };
-        });
+services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
+    .AddCookie(options =>
+    {
+        options.EventsType = typeof(CustomCookieAuthenticationEvents);
```
Comment (Tratcher) [inline aspnetcore/security/authentication/cookie.md]: I get that we're trying to show off a new 2.0 feature, but the 1.1 way to do this is still there and a lot simpler for the average case. There's no separate doc for the events model right now is there? This belongs somewhere else. @HaoK

Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: The 1.1 case conflicts with the whole point of the topic (in spirit): Cookie auth *without Identity* ... i.e., I was kind'a hoping to stay out of that namespace. I also like the new code better. As @PinpointTownes points out, this permits a good DI experience. Devs using something like this without dB access at all seems unlikely.

Comment (Tratcher) [reply aspnetcore/security/authentication/cookie.md]: ? You're using IUserRepository in both versions...

Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: Yes, but in the 1.1 case, it goes with `GetRequiredService`. I just meant that my [@]davidfowl training is that it's a bad pattern. Early on, he corrected something I did and said 'plz don't do that unless you absolutely have to.' lol

Comment (Tratcher) [reply aspnetcore/security/authentication/cookie.md]: Ah, proper DI usage is a better argument.

Comment (guardrex) [reply aspnetcore/security/authentication/cookie.md]: Yes, and getting out of the `Microsoft.AspNetCore.Identity` namespace is a *feel good moment* on this topic.

It's kind'a hard to say (in the title of the topic) that this is not using Identity ... when it really *IS*. :smile: ... even if only barely so.

**Comment by scottaddie** [comment]
@HaoK Looking for your approval on this before merging.

**Comment by HaoK** [comment]
The link to the other topic I think should be changed from "limit authorization to one instance" to => "authorize with a specific scheme"

**Comment by HaoK** [comment]
Might be worth mentioning that the `DataProtectionProvider` options property is only used to create the default `TicketDataFormat`, so if the `TicketDataFormat` property is set explicitly, the `DataProtectionProvider` is not used for anything

**Comment by HaoK** [comment]
In the validation sample, I think its worth adding throttling instead of only a warning about the performance being terrible.  You can just add another "lastChecked" claim during sign in, and then in the ValidatePrincipal method, only validate every 30 minutes instead of on every request.

**Comment by guardrex** [comment]
Thanks @HaoK 

I'm concerned about 30 minute windows on the checks: That would mean that if an employee is to be terminated, they need to be locked out in the database at least 30 minutes prior to informing them. It also means that if a hacked account is being used for malicious purposes that all users would need to be signed out with another feature (or by taking down the app for a second). I suggest we just explain that using an in-memory object for the checks would be the way to go:

> The approach described here is triggered on every request. This can result in a large performance penalty for the app. In your `ValidateLastChanged` method, you should minimize database queries as much as possible. Instead of performing a database query for every request in `ValidateLastChanged`, which would be relatively slow, you could store the database's last changed value in an in-memory object that's updated with each user database update.

There's no security risk to this approach (or legal risk to MS 😨 lol), it's fully automated, and it would be crazy faster than performing a database query to determine when the user dB was last updated.

I also just noticed that in the original version of the code samples that `ValidateLastChanged` had the `userPrincipal` supplied ... I don't think that's necessary. All we're suggesting for `ValidateLastChanged` is that it compares a `LastChanged` claim date with the last changed date for the user dB. We don't need the `userPrincipal` for that.

**Comment by Tratcher** [comment]
That doesn't work if you're app has multiple nodes. Any individual node could fall behind. The DB is the source of truth.

**Comment by guardrex** [comment]
> doesn't work if you're app has multiple nodes

Since the topic can't really address every aspect of a safe, effective setup, we'll need to leave the warning as it is. (I'll revert the change.)

**Comment by HaoK** [comment]
> I suggest we just explain that using an in-memory object for the checks would be the way to go:

So if you want to ensure the user has permissions regardless of what claims are in the request, for ultra sensitive operations you can always check the database directly (i.e. ensure the user is in the "admin" role before allowing them to delete a user)

**Comment by guardrex** [comment]
I'm still not completely convinced. If you need to fire an employee (or you lose control of a login), then you need to lock them out immediately and completely. They can't really even have read access to a system imo.

I know my former management at Pacific Charter or University Medial would *never* go for partial lockouts. Over the decades of being in business, each of these companies had to gave a number of cats the chop 🔪 under foul 😠 conditions. lol ... In one case, the employee's revenge took the form of going into the parking lot and KEYING the exec's cars!!! Yeah ... ouch. :smile:

Whatever we do with this suggestion/warning, I hope we either leave it safe with a possible perf hit for busy apps or come up with something that allows for an immediate and complete user lockout.

**Comment by guardrex** [comment]
... going back to what you, @Tratcher, said ... we could say that nodes in web farm environments would require a triggered update of their in-memory object when the user database changes. Service Broker thingy maybe?

**Comment by HaoK** [comment]
@blowdart for this his thoughts since this is how identity's signout everywhere always has worked...  we have no way to instantly invalidate all cookies, they are just flagged until the next time they are checked (which is like 20 minutes by default)  We've definitely had people ask how to force valid cookies before, but I typically tell them to check the database directly again if they really need to ensure a user has permission

**Comment by guardrex** [comment]
> check the database directly again if they really need to ensure a user has permission

I'm feeling better and better about that.

Still tho, if a `LastChecked` claim is used with an in-memory object that is updated when the dB user table is updated (including across a web farm), that should do the trick. It would be fast and safe. The devil is in the implementation of the trigger/connector for updating the nodes, but nobody said life was going to be easy.

**Comment by PinpointTownes** [comment]
> which is like 20 minutes by default

No. It's 30 minutes.

**Comment by blowdart** [comment]
Maybe we should add every request for the balance of paranoia versus performance :)

**Comment by guardrex** [comment]
@blowdart The current warning states that. Did you mean state it somewhere else in addition to the warning?

> \> [!WARNING]
> \> The approach described here is triggered on every request. This can result in a large performance penalty for the app.

**Comment by blowdart** [comment]
@HaoK can we actually set identity to do it on every request? Would that be setting to interval to zero, or does that mean never check?

**Comment by HaoK** [comment]
Nope that means always check and is basically hit it every request

**Comment by blowdart** [comment]
Oh cool.

**Comment by guardrex** [comment]
@HaoK Do we need to do anything else here?

**Comment by HaoK** [comment]
I'm happy if @blowdart is happy

**Comment by blowdart** [comment]
I'm never happy, but it's fine.

**Comment by guardrex** [comment]
Thanks guys!

@Rick-Anderson This is ready.
